ta/trusted_key/entry.c

```diff
@@ -0,0 +1,308 @@
+{
+	TEE_TASessionHandle sess = TEE_HANDLE_NULL;
+	TEE_Param params[TEE_NUM_PARAMS] = { 0 };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t ret_orig = 0;
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					       TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					       TEE_PARAM_TYPE_NONE,
+					       TEE_PARAM_TYPE_NONE);
+
+	res = TEE_OpenTASession(&(const TEE_UUID)PTA_SYSTEM_UUID,
+				TEE_TIMEOUT_INFINITE, 0, NULL, &sess,
+				&ret_orig);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	if (extra && extra_size > 0) {
+		params[0].memref.buffer = extra;
+		params[0].memref.size = extra_size;
+	}
+
+	params[1].memref.buffer = key;
+	params[1].memref.size = key_size;
+
+	res = TEE_InvokeTACommand(sess, TEE_TIMEOUT_INFINITE,
+				  PTA_SYSTEM_DERIVE_TA_UNIQUE_KEY,
+				  param_types, params, &ret_orig);
+
+	TEE_CloseTASession(sess);
+
+	return res;
+}
+
+static TEE_Result huk_AE_encrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)out;
+	uint8_t *tag = (uint8_t *)out + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)out + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = in_sz, tag_len = TAG_SIZE;
+
+	TEE_GenerateRandom(iv, IV_SIZE);
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEEncryptFinal(crypto_op, in, in_sz, enc_key, &enc_key_len,
+				 tag, &tag_len);
+	if (res != TEE_SUCCESS || tag_len != TAG_SIZE) {
+		res = TEE_ERROR_SECURITY;
+		return res;
+	}
+
+	if (ADD_OVERFLOW(enc_key_len, (IV_SIZE + TAG_SIZE), out_sz))
+		return TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_AE_decrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)in;
+	uint8_t *tag = (uint8_t *)in + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)in + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = 0;
+
+	if (SUB_OVERFLOW(in_sz, (IV_SIZE + TAG_SIZE), &enc_key_len))
+		return TEE_ERROR_SECURITY;
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEDecryptFinal(crypto_op, enc_key, enc_key_len, out, out_sz,
+				 tag, TAG_SIZE);
+	if (res != TEE_SUCCESS)
+		res = TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_crypt(TEE_OperationMode mode, void *in, size_t in_sz,
+			    void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_OperationHandle crypto_op = TEE_HANDLE_NULL;
+	TEE_ObjectHandle hkey = TEE_HANDLE_NULL;
+	uint8_t huk_key[TA_DERIVED_KEY_MAX_SIZE] = { 0 };
+	TEE_Attribute attr = { 0 };
+
+	res = TEE_AllocateOperation(&crypto_op, TEE_ALG_AES_GCM, mode,
+				    sizeof(huk_key) * 8);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = derive_unique_key(huk_key, sizeof(huk_key), NULL, 0);
+	if (res != TEE_SUCCESS) {
+		EMSG("derive_unique_key failed: returned 0x%x\n", res);
```

Comment:
`EMSG("derive_unique_key failed: returned %#"PRIx32, res);`

Mainly the 2 comments below: they also apply to other trace messages in this source file.
`res` is a u32.
no need for `\n` termination.